### PR TITLE
Configurable jitter for initial healthcheck

### PR DESCRIPTION
**Pull request by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg**
It has been observed by other parties that Envoy can send out many healthchecks more-or-less simultaneously on startup: https://github.com/envoyproxy/envoy/issues/4132. This PR fixes the issue.

I'm taking a very simple approach- let me know if this sounds reasonable:
-  The healthcheck start function- instead of calling `onIntervalBase()`, it calls `onInitialInterval()`, where, if `initial_jitter` is 0, we call `onIntervalBase()` and thus get the old behavior. If jitter is greater than zero, we instead start the interval timer.
- The jitter calculation, splitting base time + jitter into two separate functions.

I've tested this on a simple http server locally, as well adding a test to the codebase.

Fixes https://github.com/envoyproxy/envoy/issues/4132

### CODE REVIEW

**Comment by mattklein123** [comment]
@FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg thanks for adding this. Can you note in the PR description that this fixes the referenced issue? Thanks.

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
@mattklein123 added `This PR fixes the issue.` to PR description  per your comment.

File: api/envoy/api/v2/core/health_check.proto [hunk 1]
```diff
@@ -44,6 +44,10 @@ message HealthCheck {
     (gogoproto.stdduration) = true
   ];
 
+  // An optional jitter amount in milliseconds. If specified, Envoy will start after
```
Comment (snowp) [inline api/envoy/api/v2/core/health_check.proto]: Maybe `start health checking`?

File: source/common/upstream/health_checker_base_impl.cc [hunk 2]
```diff
@@ -104,14 +105,19 @@ std::chrono::milliseconds HealthCheckerImplBase::interval(HealthState state,
   } else {
     base_time_ms = no_traffic_interval_.count();
   }
+  return HealthCheckerImplBase::intervalWithJitter(base_time_ms, interval_jitter_);
```
Comment (snowp) [inline source/common/upstream/health_checker_base_impl.cc]: Do you need to qualify this with `HealthCheckerImplBase::`?

File: source/common/upstream/health_checker_base_impl.h [hunk 3]
```diff
@@ -123,6 +124,8 @@ class HealthCheckerImplBase : public HealthChecker,
   void incHealthy();
   void incDegraded();
   std::chrono::milliseconds interval(HealthState state, HealthTransition changed_state) const;
+  std::chrono::milliseconds
+  intervalWithJitter(uint64_t base_time_ms, const std::chrono::milliseconds interval_jitter) const;
```
Comment (snowp) [inline source/common/upstream/health_checker_base_impl.h]: Either make `interval_jitter` const ref or drop the `const` qualifier (see https://clang.llvm.org/extra/clang-tidy/checks/readability-avoid-const-params-in-decls.html)

Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply source/common/upstream/health_checker_base_impl.h]: ok

File: source/common/upstream/health_checker_base_impl.cc (hunk 2 above)
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply source/common/upstream/health_checker_base_impl.cc]: nope

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
@snowp updated per your comments

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
@mattklein123 apparently I have a failing macos test `//test/common/network:udp_listener_impl_test`, which is unaffected by my changes to my knowledge. Is this a known flake?

File: docs/root/intro/version_history.rst [hunk 4]
```diff
@@ -15,6 +15,7 @@ Version history
 * http: fixed a bug where large unbufferable responses were not tracked in stats and logs correctly.
 * http: fixed a crashing bug where gRPC local replies would cause segfaults when upstream access logging was on.
 * http: mitigated a race condition with the :ref:`delayed_close_timeout<envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.delayed_close_timeout>` where it could trigger while actively flushing a pending write buffer for a downstream connection.
+* initial healthcheck jitter: added :ref:`initial jitter <envoy_api_field_core.HealthCheck.initial_jitter>` to add jitter to the first healthcheck in order to prevent a thundering herd on Envoy startup.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: nit: "health check: " (and reorder) (also check to see what this is called elsewhere but I think that's the title we have used)

File: api/envoy/api/v2/core/health_check.proto [hunk 5]
```diff
@@ -44,6 +44,10 @@ message HealthCheck {
     (gogoproto.stdduration) = true
   ];
 
+  // An optional jitter amount in milliseconds. If specified, Envoy will start health
+  // checking after initial_jitter.
```
Comment (mattklein123) [inline api/envoy/api/v2/core/health_check.proto]: Can you clarify that this is a random value between 0 and initial_jitter? (Or something like that)

File: docs/root/intro/version_history.rst (hunk 4 above)
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply docs/root/intro/version_history.rst]: `health check` is used elsewhere, will fix. No reordering needed as this line is between h and j?

File: api/envoy/api/v2/core/health_check.proto (hunk 5 above)
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply api/envoy/api/v2/core/health_check.proto]: ok

File: docs/root/intro/version_history.rst (hunk 4 above)
Comment (mattklein123) [reply docs/root/intro/version_history.rst]: Technically 'e' comes before 't' :)

Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply docs/root/intro/version_history.rst]: Did you mean reorder wording of title to `health check initial jitter`, per other health check comments, and then put before `http*`?

Comment (mattklein123) [reply docs/root/intro/version_history.rst]: I mean it should read "health check: ..." like other examples.

Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply docs/root/intro/version_history.rst]: done

**Pull request merged**